drivers/axi_core/spi_engine/spi_engine.c

```diff
@@ -0,0 +1,958 @@
+				 		       mask));
+}
+
+/**
+ * @brief Add a delay bewtheen the engine commands
+ *
+ * @param desc Decriptor containing SPI interface parameters
+ * @param sleep_time_ns Number of nanoseconds to sleep between commands
+ */
+void spi_gen_sleep_ns(struct spi_desc *desc,
+		      uint32_t sleep_time_ns)
+{
+	uint32_t 		sleep_div;
+	struct spi_engine_desc	*eng_desc;
+
+	eng_desc = desc->extra;
+
+	spi_get_sleep_div(desc, sleep_time_ns, &sleep_div);
+
+	dev_dbg(desc, "Sleep for %d ns", sleep_div);
+
+	spi_engine_write_cmd_reg(eng_desc, SPI_ENGINE_CMD_SLEEP(sleep_div));
+}
+
+/**
+ * @brief Spi engine command interpreter
+ *
+ * @param desc Decriptor containing SPI interface parameters
+ * @param cmd Command to send to the engine
+ * @return int32_t - SUCCESS if the command is transfered
+ *		   - FAILURE if the command format is invalid
+ */
+int32_t spi_engine_write_cmd(struct spi_desc *desc,
+			     uint32_t cmd)
+{
+	uint8_t				engine_command;
+	uint8_t				parameter;
+	uint8_t				modifier;
+	struct spi_engine_desc		*desc_extra;
+
+	desc_extra = desc->extra;
+
+	engine_command = (cmd >> 12) & 0x0F;
+	modifier = (cmd >> 8) & 0x0F;
+	parameter = cmd & 0xFF;
+
+	switch(engine_command) {
+	case SPI_ENGINE_INST_TRANSFER:
+		dev_dbg(desc, "Transfer %d words", parameter);
+		spi_engine_transfer(desc_extra, modifier, parameter);
+		break;
+
+	case SPI_ENGINE_INST_ASSERT:
+		if(parameter == 0xFF) {
+			dev_dbg(desc, "CS HIGH");
+			/* Set the CS HIGH */
+			spi_engine_set_cs(desc, true);
+		} else if(parameter == 0x00) {
+			dev_dbg(desc, "CS LOW");
+			/* Set the CS LOW */
+			spi_engine_set_cs(desc, false);
+		}
+		break;
+
+	/* The SYNC and SLEEP commands got the same value but different
+	modifier */
+	case SPI_ENGINE_INST_SYNC_SLEEP:
+		/* SYNC instruction */
+		if(modifier == 0x00) {
+			dev_dbg(desc, "Sync ID : 0x%X", parameter);
+			spi_engine_write_cmd_reg(desc_extra, cmd);
+		} else if(modifier == 0x01) {
+			spi_gen_sleep_ns(desc, parameter);
+		}
+		break;
+	case SPI_ENGINE_INST_CONFIG:
+		switch(modifier) {
+		case 0x00:
+			dev_dbg(desc, "Prescaler set to 0x%X", parameter);
+			break;
+		case 0x01:
+			dev_dbg(desc, "SPI mode set to 0x%X", parameter);
+			break;
+		case 0x02:
+			dev_dbg(desc, "Transfer length set to 0x%X", parameter);
+			break;
+		}
+		spi_engine_write_cmd_reg(desc_extra, cmd);
+
+		break;
+
+	default:
+		dev_err(desc, "Invalid command");
+
+		return FAILURE;
+		break;
+	}
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Prepare the command queue before sending it to the engine
+ *
+ * @param desc Decriptor containing SPI interface parameters
+ * @param msg Structure used to store the transfer messages
+ * @return int32_t This function allways returns SUCCESS
+ */
+int32_t spi_engine_compile_message(struct spi_desc *desc,
+				   struct spi_engine_msg *msg)
+{
+	struct spi_engine_desc	*desc_extra;
+
+	desc_extra = desc->extra;
+
+	/* Set the data transfer length */
+	spi_engine_queue_append_cmd(&msg->cmds,
```

Comment:
`msg` needs to be tested before dereferencing

Reply:
It is an internal function. I think here msg must always come well configure.

Reply:
Well, in this case the function should be documeted accordingly, something like "this function assumes non-NULL input msg parameter". At some point during development, it is possible that an internal function becomes external, and guess what, the developer won't check the contents and how the pointers are not dereferenced, it will simply remove 'static' qualifier, and you end up with a bug.